README.md
# Nodal Background
![Nodal Background](https://raw.githubusercontent.com/david-polak/nodal-background/main/static/img/repo_background_short.jpg)
An abstract animated nodal background JavaScript canvas library.

* [Demonstration page](https://nodal-background.polakdavid.com)

## Features
* Different modes of operation
    * `AntiGravity` produces a pleasing effect of nodes pulling 
                    away from each other.
    * `Gravity` creates a true n-body simulation with emerging 
                chaotic behaviour.
    * `Simple` makes the simulation serene and non-distracting.
     
* Fully configurable
  * Everything that could have been has been exposed as a configuration 
    variable. Values can be changed that greatly affect the behaviour and 
    visuals of the simulation.
  * Concrete implementations of abstract classes can be created to achieve 
    custom behaviour.
   
* Interactive
  * Nodes can be interacted with using the mouse
  * Click scrolling changes the mass of the added node

# Usage

## Installation
[NPM Package link](https://www.npmjs.com/package/nodal-background)
```bash
npm install --save nodal-background
```

## React
### Custom wrapper
```tsx
import { useNodalBackground } from "nodal-background"

export const NodalBackgroundComponent = (props: NodalBackgroundProps) => {
  const container = useRef()
  const {setContainer} = useNodalBackground({
    container: container.current,
    ...props
  })
  
  useEffect(() => {
    if (container.current) {
      setContainer(container.current)
    }
  }, [container.current])
  
  return <div ref={container}/>
}
```

### Predefined wrapper
The wrapper above is also provided as an importable component.

```tsx
import {NodalBackgroundComponent} from "nodal-background"

const YourComponent = () => <NodalBackgroundComponent>
```

## Generic JavaScript / TypeScript
```js
new NodalBackground({container: document.getElementById("id")})
```

# Modes of operation
## Gravity Mode
https://user-images.githubusercontent.com/5686338/139601589-d5fa1313-59cd-4112-b781-e68cad123f30.mp4

### AntiGravity Mode
https://user-images.githubusercontent.com/5686338/139601633-0376d7a2-be8f-4a24-a39c-2415a0e8ce92.mp4

# Configuration

All variables can be changed while the simulation is running.

Canvas is by default transparent, changing the background color below the
canvas element combined with `nodeColor` and `linkColor` is an easy way to
achieve a different theme.

The [demonstration page](https://nodal-background.polakdavid.com) can be used
to interactively change the following controls.

## Simulation controls

#### `mode: NodalBackgroundMode`
Controls the behaviour of the entire simulation.

* `NodalBackgroundMode.AntiGravity` (default)
* `NodalBackgroundMode.Gravity`
* `NodalBackgroundMode.Simple`

#### `simMaxDistance: number, default: 150`
Maximum distance between two nodes for the interaction to be calculated.
Higher numbers are more computationally expensive and create a more connected
graph, whereas lower values produce fewer connection lines between nodes.

#### `simMinDistance: number, default: 3`
Minimum distance between two nodes for the interaction to be calculated.
Lower values produce large slingshot speeds when nodes get or spawn close 
together. Higher values minimize the slingshot potential. Too high values
will affect the perceived realism of the simulation.

#### `simMassFactor: number, default: 1`
Multiplier for node mass, affects speed and chaos.

#### `simAttraction: number, default: 10`
Multiplier for attraction, affects speed and chaos

## Node controls
#### `numberOfNodes: number, default: 100`
The target number of nodes, relative to a 1200x1200 pixel area. Simulation
complexity rises with O(n*log n), so higher numbers are discouraged.

#### `preserveNumberOfNodes: boolean, default: true`
Whether to delete extraneous nodes, with this disabled, the user created nodes
do not disappear.

#### `nodeColor: hex string, default: #000000`
Color of the nodes.

#### `nodeMaxInitialVelocity: number, default: 20`
Node initial velocity is chosen at random from a range, this control affects 
that range. Higher values significantly affect the behaviour, increasing the 
chaos.

#### `nodeInitialMass: number, default: 1.5`
Initial mass of the nodes.

#### `nodeInitialMassRange: number, default: 0`
When greater than 0, initial mass of nodes will be randomized up to the value.

#### `nodeMaxMass: number, default: 0`
Maximum mass of the nodes, applied when merging. Value `0` sets no maximum mass.

#### `nodeVisualSize: number, default: 0.5`
Visual node size offset, does not affect the simulation apart from merging.

#### `nodeAgeFactor: number, default: 0.5`
How quickly nodes fade in after spawn.

#### `nodeDeAgeFactor: number, default: 2`
How quickly nodes fade out after leaving the area.

## Link controls
#### `linkColor: hex color, default: "#000000"`
Changes the link color.

#### `linkDash: setLineDash(segments: Array<number>), default: []`
Changes the link line between nodes.

[Documentation on MDN - setLineDash()](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setLineDash)

## Performance
#### `fps: number, default: 30`
Sets the target FPS. Simulation is generally acceptable down to 20 fps.

#### `fpsCounter: boolean, defaul: false`
Displays the FPS counter, the current implementation of the fps counter has a 
negative impact on the actual fps, when bottlenecked it takes about 10fps.

## Overrides
Custom simulation behaviour can be achieved by extending the base abstract 
classes.

For more information read the included implementing classes.

#### `ticker: AbstractTicker`
Ticker performs the simulation steps. 

#### `linker: AbstractLinker`
Linker draws links between the nodes.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.3.1](https://github.com/david-polak/nodal-background/compare/v1.3.0...v1.3.1) (2022-02-05)


### Minor changes

* update homepage link ([03f68c0](https://github.com/david-polak/nodal-background/commit/03f68c04c58d616f64f66e08844f376d3a050db0))

## [1.3.0](https://github.com/david-polak/nodal-background/compare/v1.2.3...v1.3.0) (2022-02-05)


### Features

* upgrade npm dependencies ([8264a94](https://github.com/david-polak/nodal-background/commit/8264a9428a508d95a09e4cef579599e7fab75c2f))


### Minor changes

* update storybook theme and favicon ([73f6f35](https://github.com/david-polak/nodal-background/commit/73f6f3593f4da65acbfb6c29c718fb0752b0a66f))

### [1.2.3](https://github.com/david-polak/nodal-background/compare/v1.2.2...v1.2.3) (2022-02-05)


### Minor changes

* fix typo in readme ([d85a11e](https://github.com/david-polak/nodal-background/commit/d85a11e54b627685a6f409b12ce1664d15ab9e9b))

### [1.2.2](https://github.com/david-polak/nodal-background/compare/v1.2.1...v1.2.2) (2021-11-01)


### Minor changes

* make introduction the first page ([2398a51](https://github.com/david-polak/nodal-background/commit/2398a510cd8f08a8de19840472d55cdb368870e3))

### [1.2.1](https://github.com/david-polak/nodal-background/compare/v1.2.0...v1.2.1) (2021-11-01)


### Minor changes

* remove badge ([55db30c](https://github.com/david-polak/nodal-background/commit/55db30c9a70fce5e7702ffd38a4fa0f23f2c303d))

## [1.2.0](https://github.com/david-polak/nodal-background/compare/v1.1.0...v1.2.0) (2021-11-01)


### Features

* add nodeInitialMassRange ([fcc6b7b](https://github.com/david-polak/nodal-background/commit/fcc6b7b58a460bff52ec84532dc6960abf149d63))


### Minor changes

* add nodeMaxMass ([1505cfb](https://github.com/david-polak/nodal-background/commit/1505cfb3f66c662991c9ea92b9de633661680cc9))
* add npm version badge ([9db62ed](https://github.com/david-polak/nodal-background/commit/9db62edc2af89f011617f5860101e05e28c12fe3))
* update link for badge ([7ba0639](https://github.com/david-polak/nodal-background/commit/7ba0639e3f4eaebbaa89024ba427e7cad5ba89c6))

## [1.1.0](https://github.com/david-polak/nodal-background/compare/v1.0.1...v1.1.0) (2021-11-01)


### Features

* add nodeMaxMass control ([eab0352](https://github.com/david-polak/nodal-background/commit/eab035269207b86bf9ce08054e2620ee676d0516))


### Technical debt

* change the dark preset ([5093e63](https://github.com/david-polak/nodal-background/commit/5093e63208f411646690b381fb6c25091bcb0323))
* rename GravityChill files ([2477ad3](https://github.com/david-polak/nodal-background/commit/2477ad3396c691fff8463f92a78e1ebcee81ede2))


### Minor changes

* add AntiGravity Chill ([8119904](https://github.com/david-polak/nodal-background/commit/81199045ca5f153e1a4ebd78b075dfd91d684e1e))
* add Gravity Spaz ([b32f3a7](https://github.com/david-polak/nodal-background/commit/b32f3a7cd5d447d78a4346e1121a7cf0d7a15799))
* add keywords to package.json ([eb8cadd](https://github.com/david-polak/nodal-background/commit/eb8cadd06c20d6d34d9716881afc466399a0467c))
* add SimpleDash ([df73f5f](https://github.com/david-polak/nodal-background/commit/df73f5f50cccec780964c55da1a36b3315eedb28))
* fix mode in Dark theme ([d078961](https://github.com/david-polak/nodal-background/commit/d078961a54c22c9b3a27b4a3da47e6194e04dab7))
* initial set of numberOfNodes ([c040394](https://github.com/david-polak/nodal-background/commit/c04039495a933f5c5e8122d45afd5722be2cf239))

### [1.0.1](https://github.com/david-polak/nodal-background/compare/v1.0.0...v1.0.1) (2021-11-01)


### Minor changes

* add the GravityChill preset ([6d987ee](https://github.com/david-polak/nodal-background/commit/6d987ee914de63e4dddd6e6acbcf3ce1a3dc446a))
* language pass on README ([66ebbb4](https://github.com/david-polak/nodal-background/commit/66ebbb4222d11bc8568a7f15a6ab4108a7049600))

## [1.0.0](https://github.com/david-polak/nodal-background/compare/v0.1.1...v1.0.0) (2021-11-01)


### ⚠ BREAKING CHANGES

* Release 1.0.0

### Features

* Release 1.0.0 ([06040fa](https://github.com/david-polak/nodal-background/commit/06040fa91788ffd68b9987285017c1f8580b332c))

### [0.1.1](https://github.com/david-polak/nodal-background/compare/v0.1.0...v0.1.1) (2021-11-01)


### Minor changes

* fix image width in storybook ([37f55f6](https://github.com/david-polak/nodal-background/commit/37f55f66949cf3f690bc3dd8b871166871afda4f))

## [0.1.0](https://github.com/david-polak/nodal-background/compare/v0.0.10...v0.1.0) (2021-11-01)


### ⚠ BREAKING CHANGES

* prime the npm publish script

### Features

* prime the npm publish script ([74fb2e5](https://github.com/david-polak/nodal-background/commit/74fb2e5fa5d5656b974204d0cb0084bc852e65e2))


### Minor changes

* update logo in README ([65bd11f](https://github.com/david-polak/nodal-background/commit/65bd11f31493b4f35f5d2cfd7d247115a1e64575))

### [0.0.10](https://github.com/david-polak/nodal-background/compare/v0.0.9...v0.0.10) (2021-11-01)


### Features

* add publish script ([6e34624](https://github.com/david-polak/nodal-background/commit/6e346240062e0147455f9d8978c353104fb46f25))


### Minor changes

* fix typo ([f9e4e49](https://github.com/david-polak/nodal-background/commit/f9e4e499c8a8d05f3c8378e84f1676430355a49a))

### [0.0.9](https://github.com/david-polak/nodal-background/compare/v0.0.8...v0.0.9) (2021-11-01)


### Features

* add readme to storybook ([6548529](https://github.com/david-polak/nodal-background/commit/65485293c35f829ae1ece2f9c1820c368edc31b3))
* add repository background image ([334f80e](https://github.com/david-polak/nodal-background/commit/334f80edcdef5ada7cf0704dd025c825235438cf))
* update README ([1e988d5](https://github.com/david-polak/nodal-background/commit/1e988d5f1b56ff195d31b239bf21628d7ecfb4e0))


### Technical debt


src/NodalBackground.ts
import { BasicNode } from "./nodes/BasicNode"
import { AbstractNode } from "./nodes/AbstractNode"
import { BasicTicker } from "./tickers/BasicTicker"
import {
  AbstractTicker,
  InstantiableAbstractTicker,
} from "./tickers/AbstractTicker"
import {
  AbstractLinker,
  InstantiableAbstractLinker,
} from "./linkers/AbstractLinker"
import { StandardLinker } from "./linkers/StandardLinker"
import Vector2 from "./Vector2"
import { MouseHandler } from "./MouseHandler"
import { EulerTicker } from "./tickers/EulerTicker"
import { FPSCounter } from "./FPSCounter"
import { AntiEulerTicker } from "./tickers/AntiEulerTicker"

export enum NodalBackgroundMode {
  Gravity = "Gravity",
  AntiGravity = "AntiGravity",
  Simple = "Simple",
}

export interface NodalBackgroundProps {
  container: Element

  mode?: NodalBackgroundMode
  numberOfNodes?: number
  preserveNumberOfNodes: boolean

  linkColor?: string
  linkDash: Array<number>

  nodeColor?: string
  nodeMaxInitialVelocity: number
  nodeInitialMass: number
  nodeAgeFactor: number
  nodeDeAgeFactor: number
  nodeVisualSize: number

  fps?: number
  fpsCounter?: boolean

  simMaxDistance?: number
  simMinDistance?: number
  simMassFactor?: number
  simAttraction?: number

  ticker?: typeof AbstractTicker
  linker?: typeof AbstractLinker
}

export const defaultNodalBackgroundProps: NodalBackgroundProps = {
  container: null,

  mode: NodalBackgroundMode.Gravity,
  numberOfNodes: 100,
  preserveNumberOfNodes: true,

  linkColor: "#000000",
  linkDash: [],

  nodeColor: "#000000",
  nodeMaxInitialVelocity: 20,
  nodeInitialMass: 1.5,
  nodeAgeFactor: 0.5,
  nodeDeAgeFactor: 2,
  nodeVisualSize: 0.5,

  fps: 30,
  fpsCounter: false,
  ticker: null,
  linker: null,

  simMaxDistance: 150,
  simMinDistance: 3,
  simMassFactor: 1,
  simAttraction: 10,
}

export class NodalBackground {
  public props: NodalBackgroundProps

  public canvas: HTMLCanvasElement
  protected _context: CanvasRenderingContext2D

  protected _resizeListener: any
  protected _alive: boolean

  protected _ticker: AbstractTicker
  protected _linker: AbstractLinker
  protected _fpsCounter: FPSCounter
  protected _mouseHandler: MouseHandler

  protected _tFps: number
  protected _tPrevious: number

  protected _nodes: Array<AbstractNode> = []
  protected _nodesToAdd: Array<AbstractNode> = []
  protected _nodesToRemove: Array<AbstractNode> = []
  protected _nodesToMerge: Array<Array<AbstractNode>> = []
  protected _adjustedNumberOfNodes: number

  drop_distance: number

  factors: Array<Array<number | boolean>> = []

  constructor(props?: NodalBackgroundProps) {
    console.log(props)

    this.props = { ...defaultNodalBackgroundProps }

    this.props.container = props.container
    this.canvas = document.createElement("canvas")
    this.props.container.appendChild(this.canvas)

    this._context = this.canvas.getContext("2d")

    this._resizeListener = this.resize.bind(this)
    window.addEventListener("resize", this._resizeListener)

    this.nodeColor = props.nodeColor ? props.nodeColor : this.props.nodeColor
    this.nodeMaxInitialVelocity = props.nodeMaxInitialVelocity
      ? props.nodeMaxInitialVelocity
      : this.props.nodeMaxInitialVelocity
    this.nodeInitialMass = props.nodeInitialMass
      ? props.nodeInitialMass
      : this.props.nodeInitialMass
    this.nodeAgeFactor = props.nodeAgeFactor
      ? props.nodeAgeFactor
      : this.props.nodeAgeFactor
    this.nodeDeAgeFactor = props.nodeDeAgeFactor
      ? props.nodeDeAgeFactor
      : this.props.nodeDeAgeFactor
    this.nodeVisualSize = props.nodeVisualSize
      ? props.nodeVisualSize
      : this.props.nodeVisualSize

    this.resize()

    this.mode = props.mode ? props.mode : this.props.mode
    this.ticker = props.ticker ? props.ticker : this.props.ticker
    this.simMaxDistance = props.simMaxDistance
      ? props.simMaxDistance
      : this.props.simMaxDistance
    this.simMinDistance = props.simMinDistance
      ? props.simMinDistance
      : this.props.simMinDistance
    this.simMassFactor = props.simMassFactor
      ? props.simMassFactor
      : this.props.simMassFactor
    this.simAttraction = props.simAttraction
      ? props.simAttraction
      : this.props.simAttraction

    this.linker = props.linker ? props.linker : this.props.linker
    this.linkColor = props.linkColor ? props.linkColor : this.props.linkColor
    this.linkDash = props.linkDash ? props.linkDash : this.props.linkDash

    this.numberOfNodes = props.numberOfNodes
      ? props.numberOfNodes
      : this.props.numberOfNodes
    this.preserveNumberOfNodes = props.preserveNumberOfNodes
      ? props.preserveNumberOfNodes
      : this.preserveNumberOfNodes

    this.fps = props.fps ? props.fps : this.props.fps
    this.fpsCounter = props.fpsCounter
      ? props.fpsCounter
      : this.props.fpsCounter

    this.drop_distance = 0
    this._mouseHandler = new MouseHandler(this.canvas, this.addNode.bind(this))

    this._tPrevious = Date.now()
    this._alive = true
    requestAnimationFrame(this.handleAnimationFrame.bind(this))
  }

  set linkColor(linkColor: string) {
    this.props.linkColor = linkColor
    this._linker.linkColor = linkColor
  }

  set nodeColor(nodeColor: string) {
    this.props.nodeColor = nodeColor
    this._nodes.forEach((node) => (node.nodeColor = nodeColor))
  }

  set nodeMaxInitialVelocity(nodeMaxInitialVelocity: number) {
    this.props.nodeMaxInitialVelocity = nodeMaxInitialVelocity
  }

  set nodeInitialMass(nodeInitialMass: number) {
    this.props.nodeInitialMass = nodeInitialMass
  }

  set nodeAgeFactor(nodeAgeFactor: number) {
    this.props.nodeAgeFactor = nodeAgeFactor
    this._nodes.forEach((node) => (node.ageFactor = nodeAgeFactor))
  }

  set nodeDeAgeFactor(nodeDeAgeFactor: number) {
    this.props.nodeDeAgeFactor = nodeDeAgeFactor
    this._nodes.forEach((node) => (node.deAgeFactor = nodeDeAgeFactor))
  }

  set nodeVisualSize(nodeVisualSize: number) {
    this.props.nodeVisualSize = nodeVisualSize
    this._nodes.forEach((node) => (node.visualSize = nodeVisualSize))
  }

  set ticker(ticker: typeof AbstractTicker) {
    if (!ticker) {
      return
    }
    this.props.ticker = ticker
    const instantiable = ticker as InstantiableAbstractTicker<AbstractTicker>
    this._ticker = new instantiable()
  }

  set linker(linker: typeof AbstractLinker) {
    if (!linker) {
      this._linker = new StandardLinker(this._context)
      return
    }
    this.props.linker = linker
    const instantiable = linker as InstantiableAbstractLinker<AbstractLinker>
    this._linker = new instantiable(this._context)
  }

  set linkDash(linkDash: Array<number>) {
    this.props.linkDash = linkDash
    this._linker.linkDash = linkDash
  }

  set mode(mode: NodalBackgroundMode) {
    this.props.mode = mode

    if (mode === NodalBackgroundMode.Gravity) {
      this._ticker = new EulerTicker()
    } else if (mode === NodalBackgroundMode.AntiGravity) {
      this._ticker = new AntiEulerTicker()
    } else if (mode === NodalBackgroundMode.Simple) {
      this._ticker = new BasicTicker()
    } else {
      this._ticker = new EulerTicker()
    }

    this._ticker.maxDistance = this.props.simMaxDistance
    this._ticker.minDistance = this.props.simMinDistance
    this._ticker.massFactor = this.props.simMassFactor
    this._ticker.attraction = this.props.simAttraction
  }

  set simMaxDistance(simMaxDistance: number) {
    this.props.simMaxDistance = simMaxDistance
    this._ticker.maxDistance = simMaxDistance
  }
  set simMinDistance(simMinDistance: number) {
    this.props.simMinDistance = simMinDistance
    this._ticker.minDistance = simMinDistance
  }

  set simMassFactor(simMassFactor: number) {
    this.props.simMassFactor = simMassFactor
    this._ticker.massFactor = simMassFactor
  }

  set simAttraction(simAttraction: number) {
    this.props.simAttraction = simAttraction
    this._ticker.attraction = simAttraction
  }

  set numberOfNodes(numberOfNodes: number) {
    this.props.numberOfNodes = numberOfNodes

    const pixels = this.canvas.width * this.canvas.height
    const pixelTarget = 1200 * 1200
    this._adjustedNumberOfNodes = (pixels / pixelTarget) * numberOfNodes

    while (
      this._adjustedNumberOfNodes >
      this._nodes.length + this._nodesToAdd.length
    ) {
      this.addNode()
    }
  }

  set fps(fps: number) {
    this.props.fps = fps
    this._tFps = (1 / fps) * 1000
  }

  set fpsCounter(fpsCounter: boolean) {
    this.props.fpsCounter = fpsCounter
    this._fpsCounter = new FPSCounter(this._context, fpsCounter)
  }

  set preserveNumberOfNodes(preserveNumberOfNodes: boolean) {
    this.props.preserveNumberOfNodes = preserveNumberOfNodes
  }

  destroy() {
    if (!this._alive) {
      this._alive = false
      return
    }
    // The rest of this method will execute in the next animation frame.

    window.removeEventListener("resize", this._resizeListener)
    this.canvas.remove()
    console.log("TODO: FINISH CLEANUP!")
  }

  protected resize() {
    this.canvas.width = this.props.container.clientWidth * devicePixelRatio
    this.canvas.height = this.props.container.clientHeight * devicePixelRatio

    // recalculate the _adjustedNumberOfNodes
    this.numberOfNodes = this.props.numberOfNodes
  }

  handleAnimationFrame() {
    if (!this._alive) {
      this.destroy()
      return
    }

    const now = Date.now()
    const tDelta = now - this._tPrevious

    if (tDelta > this._tFps) {
      this.tick(tDelta)
      this._tPrevious = now - (tDelta % this._tFps)
    }

    requestAnimationFrame(this.handleAnimationFrame.bind(this))
  }

  addNode(): AbstractNode {
    const node = new BasicNode(
      this.canvas,
      this.props.nodeMaxInitialVelocity,
      this.props.nodeInitialMass
    )
    node.nodeColor = this.props.nodeColor
    node.visualSize = this.props.nodeVisualSize
    node.ageFactor = this.props.nodeAgeFactor
    node.deAgeFactor = this.props.nodeDeAgeFactor
    this._nodesToAdd.push(node)
    return node
  }

  mergeNodes(nodeA: AbstractNode, nodeB: AbstractNode) {
    // conservation of momentum
    nodeA.velocity.x =
      (nodeA.mass * nodeA.velocity.x + nodeB.mass * nodeB.velocity.x) /
      (nodeA.mass + nodeB.mass)
    nodeA.velocity.y =
      (nodeA.mass * nodeA.velocity.y + nodeB.mass * nodeB.velocity.y) /
      (nodeA.mass + nodeB.mass)
    nodeA.mass += nodeB.mass

    nodeB.recreate(
      this.props.nodeMaxInitialVelocity,
      this.props.nodeInitialMass
    )
  }

  tick(time: number) {
    /*
     * This tick method should come out to O(nlogn), there are alternative
     * paths which would sacrifice rendering precision at lower fps for less
     * operations per tick.
     */

    this._context.clearRect(0, 0, this.canvas.width, this.canvas.height)

    this._fpsCounter.draw(time)

    while (this._nodesToRemove.length) {
      const node = this._nodesToRemove.pop()
      const index = this._nodes.findIndex((n) => n === node)
      this._nodes.splice(index, 1)
    }

    while (this._nodesToAdd.length) {
      this._nodes.push(this._nodesToAdd.pop())
      this.factors.push([])
    }

    for (let i = 0; i < this._nodes.length; i++) {
      for (let j = i + 1; j < this._nodes.length; j++) {
        // simulation step for both nodes
        this.factors[i][j] = this._ticker.tickBoth(
          time,
          this._nodes[i],
          this._nodes[j]
        )
      }
    }

    for (let i = 0; i < this._nodes.length; i++) {
      const node: AbstractNode = this._nodes[i]

      // simulation step for single node
      this._ticker.tickSingle(time, node)

      // resetting the mouse node's position
      this._mouseHandler.tick()
    }

    for (let i = 0; i < this._nodes.length; i++) {
      for (let j = i + 1; j < this._nodes.length; j++) {
        let factor = this.factors[i][j]

        if (factor === true || factor === false) {
          this._nodesToMerge.push([this._nodes[i], this._nodes[j]])
          factor = 0
        }

        if (factor > 0.01) {
          this._linker.renderLink(factor, this._nodes[i], this._nodes[j])
        }
      }
    }

    while (this._nodesToMerge.length) {
      const toMerge: Array<AbstractNode> = this._nodesToMerge.pop()
      const nodeA: AbstractNode = toMerge[0]
      const nodeB: AbstractNode = toMerge[1]
      if (nodeA.mass > nodeB.mass) {
        this.mergeNodes(nodeA, nodeB)
      } else {
        this.mergeNodes(nodeB, nodeA)
      }
    }

    for (let i = 0; i < this._nodes.length; i++) {
      const nodeA: AbstractNode = this._nodes[i]

      nodeA.render()

      if (
        nodeA.position.x < -this.drop_distance ||
        nodeA.position.y < -this.drop_distance ||
        nodeA.position.x > this.canvas.width + this.drop_distance ||
        nodeA.position.y > this.canvas.height + this.drop_distance
      ) {
        if (nodeA.age < 0) {
          if (!this.props.preserveNumberOfNodes) {
            nodeA.recreate(
              this.props.nodeMaxInitialVelocity,
              this.props.nodeInitialMass
            )
            continue
          }

          if (
            this._nodes.length + this._nodesToRemove.length >
            this._adjustedNumberOfNodes
          ) {
            this._nodesToRemove.push(nodeA)
          } else {
            nodeA.recreate(
              this.props.nodeMaxInitialVelocity,
              this.props.nodeInitialMass
            )
          }
        } else if (nodeA.age > 1) {
          nodeA.age = 1
        } else {
          nodeA.deAgeNode(time)
        }
      } else {
        nodeA.ageNode(time)
      }
    }
  }
}
